### PR TITLE
Fix unable to populate User in Users-Permissions

### LINKED ISSUE #11957: Population does not work for Users in Users-Permissions
## Bug report

When we try to populate the relational fields on the user-permissions => users model, the response doesn't populate the relational fields.

### Steps to reproduce the behavior

1. Create a user in the Users-Permissions
2. Create some other collection type and link it with user
3. Enable public find permission on both
4. Send a GET request to the /users with "?populate=*"
5. Response won't contain the relational field

### Expected behavior

Appending ?populate=* to the API request to the user-permissions => users model, the response should contain the relational fields.

### System

- Node.js version: 14.15.3
- NPM version: 6.14.9
- Strapi version: 4.0.0
- Database: MariaDB
- Operating system: MacOS

### Additional context

PR  #11795 is about fixing filtering on the same User-permissions => users model. I have also added that new code changes to my project. Filtering does work, but the population doesn't.

### PR DESCRIPTION
**Pull request by iicdii**
Signed-off-by: harimkims <harimkims@gmail.com>

### What does it do?

- Fix unable to populate User/Users in Users-Permissions plugin

### Why is it needed?

- To enable population for user/users

### How to test it?

1. Create user
2. Send GET request `http://localhost:1337/api/users?populate=*`
3. Send GET request `http://localhost:1337/api/users/<userId>?populate=*`

### Related issue(s)/PR(s)

Closes #11957

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
@iicdii you are amazing, I hope you know that. :hugs:

**Comment by derrickmehaffy** [comment]
Alex/others I've added this to the v4.0.1 milestone just in hopes of pushing it through on release as we are getting -a lot- of questions about this.

If it's not possible feel free to remove.

**Comment by derrickmehaffy** [comment]
This pull request has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/populate-users-permissions-to-get-user-avatar/13701/2

**Comment by iicdii** [comment]
Thanks! 😊 Yeah I've already seen many questions about this issue on Discord too. Hope it could be merged into v4.0.1!

**Comment by derrickmehaffy** [comment]
This pull request has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/strapi-v4-no-relation-fields-in-user-collection-types/13226/8

**Review by Convly — COMMENTED** [review]
Thanks for opening yet another PR! :raised_hands: 

Tbh, I don't think this logic belongs to the REST controller. Even if for now the GQL resolvers are directly calling UP controllers (it's temporary), the goal is to make them both call a common service where we could put common transformations etc...

Also, I'm kinda curious about why we're not using the entity service API to fetch the entries in the UP's user service. Since the `convertPopulate` & other transformers would be called automatically.

@alexandrebodin do you remember why this choice has been made? Is there a particular reason? Or could we replace those queries with entity service's ones?

**tl;dr: this logic is already handled by the entity service, maybe we can make the UP user service use it instead of doing DB queries directly + it would create a common API for both REST controllers & GraphQL resolvers.**

**Comment by iicdii** [comment]
Actually I've tried to convert them to entityService in previous PR, He said it like below which is the reason why I don't use entityService
> I'm not sure this is something we want to move forward with as it only changes the find but doesn't change the findOne etc.
>
>This API is not like the autogenerated ones by design and only supports filters passed at the root level of the query.
>You are not supposed to be able to pass down sorting or limit for example.
>
>I agree it would be better to make it consistent but we have some security concerns to take into account first.
>https://github.com/strapi/strapi/pull/11795#issuecomment-993344305

**Comment by derrickmehaffy** [comment]
@iicdii / @Convly I also don't know why we don't move to ES, the U&P plugin is quite a fews behind on rewrites to get it anywhere close to the other plugins.

**Comment by Convly** [comment]
Then we'll keep it that way and convert everything to ES later. But we'll definitely have to take care of this ASAP imo.

**Comment by Convly** [comment]
Could you update your branch with master, please?

**Comment by Convly** [comment]
> @iicdii / @Convly I also don't know why we don't move to ES, the U&P plugin is quite a fews behind on rewrites to get it anywhere close to the other plugins.

Definitely a topic we'll need to tackle as soon as we have some bandwidth!

**Comment by derrickmehaffy** [comment]
> > @iicdii / @Convly I also don't know why we don't move to ES, the U&P plugin is quite a fews behind on rewrites to get it anywhere close to the other plugins.
> 
> Definitely a topic we'll need to tackle as soon as we have some bandwidth!

Ah might have gotten skipped in the rush for v4 launch. Can take that later on aye

**Comment by alexandrebodin** [comment]
FYI this API never was meant to have the same features as the auto-generated content APIs. 

There are some security concerns around users and we just cannot expose a complete populate option without more considerations 1st. I'll flag the PR so we can review the logic thoroughly before moving forward

**Comment by derrickmehaffy** [comment]
This pull request has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/strapi-v4-no-relation-fields-in-user-collection-types/13226/12

**Comment by M1CK431** [comment]
Hi,

This issue is really a big show stopper for me (and probably for almost everyone trying to build an authenticated API). I'm currently migrating from v3 to v4 and my release is planned in about 1 month.

Sorry to ask that, no stress, but could I reasonably expect a (released) fix before my deadline (aka mid-feb)?

NB: thanks a lot @iicdii for this PR :kissing_heart:

**Comment by derrickmehaffy** [comment]
> FYI this API never was meant to have the same features as the auto-generated content APIs.
> 
> There are some security concerns around users and we just cannot expose a complete populate option without more considerations 1st. I'll flag the PR so we can review the logic thoroughly before moving forward

@alexandrebodin if the security concerns are about population of the role and getting other users information couldn't we just change the user <> role relation from oneToOne to oneWay (or even better manyWay) and be done with it?

The deepest someone could populate would be user => role => permissions and there shouldn't be any risk to getting other users information.

**Comment by hblab-hautp** [comment]
Hi team !
My work was delayed because of this issue. 
A micro service should have an authenticated API , so at least we have to get user's role.
Beside it, an application for multiple user group need apply multi tenant, the simplest way is using a same database 

Sorry for my asking, but i hope it's going to be resolved  as soon as possible !

Thank you, love u al

**Comment by lucasm107** [comment]
> Hi,
> 
> This issue is really a big show stopper for me (and probably for almost everyone trying to build an authenticated API). I'm currently migrating from v3 to v4 and my release is planned in about 1 month.
> 
> Sorry to ask that, no stress, but could I reasonably expect a (released) fix before my deadline (aka mid-feb)?
> 
> NB: thanks a lot @iicdii for this PR 😘

+1 

Someone have information if this is going to be in some future version? because the "don't merge" label feels that this can be rejected ?

**Comment by derrickmehaffy** [comment]
> > Hi,
> > This issue is really a big show stopper for me (and probably for almost everyone trying to build an authenticated API). I'm currently migrating from v3 to v4 and my release is planned in about 1 month.
> > Sorry to ask that, no stress, but could I reasonably expect a (released) fix before my deadline (aka mid-feb)?
> > NB: thanks a lot @iicdii for this PR kissing_heart
> 
> +1
> 
> Someone have information if this is going to be in some future version? because the "don't merge" label feels that this can be rejected ?

We will fix the issue, I believe this is waiting on the backend engineering team, specifically @alexandrebodin to validate the fix due it's security nature.

**Comment by MostafaHassan** [comment]
It looks like it works.
However, it doesn't seem to work for when the user gets authenticated. 
Example, I would like to something like this as well: /api/auth/local?populate=*

**Comment by hblab-hautp** [comment]
> It looks like it works. However, it doesn't seem to work for when the user gets authenticated. Example, I would like to do something like this as well: /api/auth/local?populate=*

No, it's not.
I'm using strapi v4 and i can not get any populate when try to call "/api/auth/local?populate=*", populate is not work.
I just need the user's role to auth my account at my React App

Have any way to fix it now ? please tell me if you found
@MostafaHassan

**Comment by MostafaHassan** [comment]
> > It looks like it works. However, it doesn't seem to work for when the user gets authenticated. Example, I would like to do something like this as well: /api/auth/local?populate=*
> 
> No, it's not. I'm using strapi v4 and i can not get any populate when try to call "/api/auth/local?populate=*", populate is not work. I just need the user's role to auth my account at my React App
> 
> Have any way to fix it now ? please tell me if you found @MostafaHassan

I meant the code in this PR seems to work for populaing /api/users/me and /api/users/:id 
And there is another way of making populate work for those two endpoints. But I have no idea how to get populate to work for /api/auth/local?populate=*

Yes, I know that neither of these is working on master.

**Comment by toekneema** [comment]
any updates?

**Comment by derrickmehaffy** [comment]
> > FYI this API never was meant to have the same features as the auto-generated content APIs.
> > There are some security concerns around users and we just cannot expose a complete populate option without more considerations 1st. I'll flag the PR so we can review the logic thoroughly before moving forward
> 
> @alexandrebodin if the security concerns are about population of the role and getting other users information couldn't we just change the user <> role relation from oneToOne to oneWay (or even better manyWay) and be done with it?
> 
> The deepest someone could populate would be user => role => permissions and there shouldn't be any risk to getting other users information.
> 
> CC @JabStrapi

@kasonde and I were playing the other day on swapping `oneToOne` for `oneWay` and nothing broke that we could tell. So this should be a valid option to fix the security risk of population. Assuming the user build their own relations with that in mind also

**Comment by be90728** [comment]
Hope this can help you:

Add folder like this
`src/extentions/users-permissions`
Also create `strapi-server.js` to users-permissions folder

For example, if i want populate `role` in result without query params:
`strapi-server.js`
`
module.exports = plugin => {
    const sanitizeOutput = (user) => {
        const {
            password, resetPasswordToken, confirmationToken, ...sanitizedUser
        } = user; // be careful, you need to omit other private attributes yourself
        return sanitizedUser;
    };

    plugin.controllers.user.me = async (ctx) => {
        if (!ctx.state.user) {
            return ctx.unauthorized();
        }
        const user = await strapi.entityService.findOne(
            'plugin::users-permissions.user',
            ctx.state.user.id,
            { populate: ['role'] }
        );

        ctx.body = sanitizeOutput(user);
    };

    plugin.controllers.user.find = async (ctx) => {
        const users = await strapi.entityService.findMany(
            'plugin::users-permissions.user',
            { ...ctx.params, populate: ['role'] }
        );

        ctx.body = users.map(user => sanitizeOutput(user));
    };

    plugin.controllers.user.findOne = async (ctx) => {
        if (!ctx.state.user) {
            return ctx.unauthorized();
        }
        const user = await strapi.entityService.findOne(
            'plugin::users-permissions.user',
            ctx.state.user.id,
            { populate: ['role'] }
        );

        ctx.body = sanitizeOutput(user);
    };

    return plugin;
};
`

**Comment by be90728** [comment]
Hope this can help you:

Add folder like this
`project-folder/src/extentions/users-permissions`
Also create `strapi-server.js` to users-permissions folder

For example, if i want populate `role` in result without query params:

```
// project-folder/src/extentions/users-permissions/strapi-server.js

module.exports = plugin => {
    const sanitizeOutput = (user) => {
        const {
            password, resetPasswordToken, confirmationToken, ...sanitizedUser
        } = user; // be careful, you need to omit other private attributes yourself
        return sanitizedUser;
    };

    plugin.controllers.user.me = async (ctx) => {
        if (!ctx.state.user) {
            return ctx.unauthorized();
        }
        const user = await strapi.entityService.findOne(
            'plugin::users-permissions.user',
            ctx.state.user.id,
            { populate: ['role'] }
        );

        ctx.body = sanitizeOutput(user);
    };

    plugin.controllers.user.find = async (ctx) => {
        const users = await strapi.entityService.findMany(
            'plugin::users-permissions.user',
            { ...ctx.params, populate: ['role'] }
        );

        ctx.body = users.map(user => sanitizeOutput(user));
    };

    plugin.controllers.user.findOne = async (ctx) => {
        if (!ctx.state.user) {
            return ctx.unauthorized();
        }
        const user = await strapi.entityService.findOne(
            'plugin::users-permissions.user',
            ctx.state.user.id,
            { populate: ['role'] }
        );

        ctx.body = sanitizeOutput(user);
    };

    return plugin;
};
```

**Comment by netchart** [comment]
> Hope this can help you:
> 
> Add folder like this `project-folder/src/extentions/users-permissions` Also create `strapi-server.js` to users-permissions folder
> 
> For example, if i want populate `role` in result without query params:
> 
> ```
> // project-folder/src/extentions/users-permissions/strapi-server.js
> 
> module.exports = plugin => {
>     const sanitizeOutput = (user) => {
> ...
>     return plugin;
> };
> ```

Excelent!, it worked for me, but it is ../extensions/.. right? (instead of ../extentions/..)

**Comment by derrickmehaffy** [comment]
@iicdii I spoke with @alexandrebodin and @Convly today in a meeting to figure out how best to handle this (among other U&P errors).

Are you willing to rewrite this plugin to match up with the entityService assuming we can help figure out how to sanitize the nested user lookup? @alexandrebodin seems to believe that changing the user <--> role relation from manyToOne to oneWay would be a breaking change but I don't believe so.

Happy to talk on Discord if you want just PM me. :)

**Comment by alexandrebodin** [comment]
> @iicdii I spoke with @alexandrebodin and @Convly today in a meeting to figure out how best to handle this (among other U&P errors).
> 
> Are you willing to rewrite this plugin to match up with the entityService assuming we can help figure out how to sanitize the nested user lookup? @alexandrebodin seems to believe that changing the user <--> role relation from manyToOne to oneWay would be a breaking change but I don't believe so.
> 

> Happy to talk on Discord if you want just PM me. :)

**Comment by RiusmaX** [comment]
Hi, 
Any updates ? I can't get the "full" user object from API (strapi v4.1.0).
I've a nested component and a media, there is no way to get it with the REST API but it's working with GraphQL.

**Comment by alexandrebodin** [comment]
Hi @RiusmaX, sadly no. At the moment this is being discussed and we are lacking bandwidth to work on it. 

However, if you really need this asap, you can fairly easily create a custom API endpoint that will populate what you want. Let us know if you don't see how to do it and we will try to help out as best.

**Comment by itai-rozen** [comment]
So as of now i see the problem is not fixed yet.. I saw you suggested  to create a custom API endpoint that will populate whatevet i want, but im really new to strapi and not sure what you meant:)
any chance you could guide me?

**Review by Convly — CHANGES_REQUESTED** [review]
Hey, thanks for the update.
As stated in my comment on `packages/core/utils/lib/sanitize/index.js`, we shouldn't couple the UP plugin with Strapi core.
Ideally, it would make more sense to move the logic [over there](https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/controllers/user.js#L17-L22). However it would mean that this transformation/sanitation wouldn't be applied to any relation made to the UP users or role.
Maybe we're missing a way to extend the sanitation process. @alexandrebodin what would you think of adding a way to add transforms (sanitation steps) to the default `sanitizeOutput` & `sanitizeInput` used in the core controllers? :thinking:

**Comment by georgeiliadis91** [comment]
How is this not merged yet?  Not being able to have a profile img is a deal breaker.

**Comment by forntoh** [comment]
> > So as of now i see the problem is not fixed yet.. I saw you suggested to create a custom API endpoint that will populate whatevet i want, but im really new to strapi and not sure what you meant:) any chance you could guide me?
> 
> Hi, in Strapi you can change the code to match what you want. I found some informations here : https://forum.strapi.io/t/create-custom-endpoint-for-users-in-strapi-v4/13724

I wish to know if this also works with the graphql plugin,
I am trying to get the users role in `usersPermissionsUser(id: $userId)`

**Comment by iicdii** [comment]
@Convly I added sanitizers registry in core and moved the sanitization to U&P register function from core.

**Comment by derrickmehaffy** [comment]
Have been testing this extensively over the past few days and it's working really well! For anyone else who wants to test (wouldn't recommend anyone deploy to production yet) here are my `patch-package` patch files:

[patches.zip](https://github.com/strapi/strapi/files/8550674/patches.zip)

These patches were generated on Strapi v4.1.8

**Comment by forntoh** [comment]
> Have been testing this extensively over the past few days and it's working really well! For anyone else who wants to test (wouldn't recommend anyone deploy to production yet) here are my `patch-package` patch files:
> 
> [patches.zip](https://github.com/strapi/strapi/files/8550674/patches.zip)
> 
> These patches were generated on Strapi v4.1.8

Thank you so much, I love the progress that your team is making, at this rate, v5 will be out of this world
How do I use/test the patch files?

**Comment by derrickmehaffy** [comment]
> > Have been testing this extensively over the past few days and it's working really well! For anyone else who wants to test (wouldn't recommend anyone deploy to production yet) here are my `patch-package` patch files:
> > [patches.zip](https://github.com/strapi/strapi/files/8550674/patches.zip)
> > These patches were generated on Strapi v4.1.8
> 
> Thank you so much, I love the progress that your team is making, at this rate, v5 will be out of this world How do I use/test the patch files?

You'll need to install patch-package (if you follow their docs it's just a matter of installing the packages, depends on if you use npm or yarn) and don't forget to add the postinstall script to your package.json. Then you can just place the scripts folder from inside the zip at the root of your project and reinstall your node_modules. The patches should get used once the install is finished.

**Comment by jonasstams** [comment]
Is there anyway to bypass this issue?

File: packages/core/strapi/lib/Strapi.js
Comment (Convly) [inline packages/core/strapi/lib/Strapi.js:165]: What do you think of adding an alias to get a specific sanitizer here? We're doing the same for services, controllers, etc...

Smth like
```javascript
  get sanitizer(uid) {
    return this.container.get('sanitizers').get(uid);
  }
```

File: packages/core/strapi/lib/core/registries/sanitizers.js
Comment (Convly) [inline packages/core/strapi/lib/core/registries/sanitizers.js:23]: What do you think of adding a way to register new sanitizers? Something/A method to add a new key somewhere in a the sanitizers object?
IMO, it could be useful for plugin creators, etc...

**Review by Convly — CHANGES_REQUESTED** [review]
Thanks so much for your commitment & your dedication to this @iicdii.
I've added some comments, tell me what you think :slightly_smiling_face:

File: packages/core/strapi/lib/core/registries/sanitizers.js
Comment (iicdii) [reply packages/core/strapi/lib/core/registries/sanitizers.js:7]: I didn't know about the loader. I'll apply it :)

File: packages/core/strapi/lib/core/registries/sanitizers.js
Comment (iicdii) [reply packages/core/strapi/lib/core/registries/sanitizers.js:25]: That looks good. I haven't thought about the use case yet, but it will help with scalability for sure.

**Review by Convly — APPROVED** [review]
LGTM, nice job, I think both the community & the Strapi team owes you a big one for this.
I'll play a bit with it before merging it. I'll try to do it fast so that it can be part of the next release tomorrow.

**Review by alexandrebodin — APPROVED** [review]
Thank you for this fix @iicdii 💯 

We will probably have to rethink the sanitizers approach soon but this is a working solution that will please a lot of users so we will merge it now and handle the core code soon.

**Comment by forntoh** [comment]
> Thank you for this fix @iicdii 💯 
> 
> 
> 
> We will probably have to rethink the sanitizers approach soon but this is a working solution that will please a lot of users so we will merge it now and handle the core code soon.

We've been waiting for this and finally it's here, @iicdii you're amazing ☺️.

**Comment by umair-me** [comment]
Thank you @iicdii 🙏

**Comment by Nikoxx99** [comment]
@iicdii Thank you! what a legend

**Comment by violabg** [comment]
it stil doesn't work on graphQl after updating to version 4.1.10

```gql
query getUsers {
  usersPermissionsUser(id: 1) {
      data {
        id
        attributes {
          firstName
          lastName
          blocked
          role{
            data{
              attributes {
                type
              }
            }
          }
        }
      }
    }
}
```

```gql
{
  "errors": [
    {
      "message": "Forbidden access",
      "extensions": {
        "error": {
          "name": "ForbiddenError",
          "message": "Forbidden access",
          "details": {}
        },
        "code": "FORBIDDEN"
      }
    }
  ],
  "data": {
    "usersPermissionsUser": {
      "data": {
        "id": "1",
        "attributes": {
          "firstName": "xxx",
          "lastName": "xxx",
          "blocked": false,
          "role": null
        }
      }
    }
  }
}

**Comment by M1CK431** [comment]
Hi @iicdii and thanks A LOT for your work! :heart: 

However I have a question: how to handle extra fields in `User` since using
```js
register({ strapi }) {
    const extensionService = strapi.plugin("graphql").service("extension");
    extensionService.use(({ nexus }) => ({
      types: [
        nexus.extendType({
          type: "UsersPermissionsMe",
          definition(t) {
            // here define fields you need
            t.string("firstName");
            t.string("lastName");
            t.field("acl", { type: "JSON" });
            t.dateTime("deletedAt");
          }
        })
      ]
    }));
  },
```
in `index.js` then trying to mutate `acl` (for example) produce this error:
```
[2022-05-12 01:26:46.519] error: update `up_users` set `blocked` = false, `acl` = '{\"userManagement\":false}', `updated_at` = '2022-05-12 01:26:46.489' where (`id` = 1) - ER_BAD_FIELD_ERROR: Unknown column 'acl' in 'field list'
Error: ER_BAD_FIELD_ERROR: Unknown column 'acl' in 'field list'
    at Query.Sequence._packetToError (/home/tuut/Projets/biip/api/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
    at Query.ErrorPacket (/home/tuut/Projets/biip/api/node_modules/mysql/lib/protocol/sequences/Query.js:79:18)
    at Protocol._parsePacket (/home/tuut/Projets/biip/api/node_modules/mysql/lib/protocol/Protocol.js:291:23)
    at Parser._parsePacket (/home/tuut/Projets/biip/api/node_modules/mysql/lib/protocol/Parser.js:433:10)
    at Parser.write (/home/tuut/Projets/biip/api/node_modules/mysql/lib/protocol/Parser.js:43:10)
    at Protocol.write (/home/tuut/Projets/biip/api/node_modules/mysql/lib/protocol/Protocol.js:38:16)
    at Socket.<anonymous> (/home/tuut/Projets/biip/api/node_modules/mysql/lib/Connection.js:88:28)
    at Socket.<anonymous> (/home/tuut/Projets/biip/api/node_modules/mysql/lib/Connection.js:526:10)
    at Socket.emit (node:events:390:28)
    at Socket.emit (node:domain:475:12)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:199:23)
    --------------------
    at Protocol._enqueue (/home/tuut/Projets/biip/api/node_modules/mysql/lib/protocol/Protocol.js:144:48)
    at Connection.query (/home/tuut/Projets/biip/api/node_modules/mysql/lib/Connection.js:198:25)
    at /home/tuut/Projets/biip/api/node_modules/knex/lib/dialects/mysql/index.js:132:18
    at new Promise (<anonymous>)
    at Client_MySQL._query (/home/tuut/Projets/biip/api/node_modules/knex/lib/dialects/mysql/index.js:126:12)
    at executeQuery (/home/tuut/Projets/biip/api/node_modules/knex/lib/execution/internal/query-executioner.js:37:17)
    at Client_MySQL.query (/home/tuut/Projets/biip/api/node_modules/knex/lib/client.js:146:12)
    at Runner.query (/home/tuut/Projets/biip/api/node_modules/knex/lib/execution/runner.js:130:36)
    at ensureConnectionCallback (/home/tuut/Projets/biip/api/node_modules/knex/lib/execution/internal/ensure-connection-callback.js:13:17)
    at Runner.ensureConnection (/home/tuut/Projets/biip/api/node_modules/knex/lib/execution/runner.js:307:20)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Runner.run (/home/tuut/Projets/biip/api/node_modules/knex/lib/execution/runner.js:30:19)
    at async Object.execute (/home/tuut/Projets/biip/api/node_modules/@strapi/database/lib/query/query-builder.js:354:22)
    at async Object.update (/home/tuut/Projets/biip/api/node_modules/@strapi/database/lib/entity-manager.js:247:9)
    at async Object.update (/home/tuut/Projets/biip/api/node_modules/@strapi/strapi/lib/services/entity-service/index.js:219:18)
```

Also, the added fields are still present in GraphQL documentation but doesn't return any data while fetched.
For now, because of that, I'm forced to revert on v4.1.9 despite I was waiting for a very long time after this issue fix :disappointed:

**Comment by msoler75** [comment]
I'm not sure if I'am doing well but even after version update to 4.1.10 I can't populate fields with the /api/users/me controller

`/api/users/me?[populate][relation]=*`

So I put this code based on some code I found somewhere here:


```
// ISSUE: 
// https://github.com/strapi/strapi/issues/11957

const Query = require("mysql/lib/protocol/sequences/Query");

// BUGFIX
module.exports = plugin => {
    const sanitizeOutput = (user) => {
      const {
        password,
        resetPasswordToken,
        confirmationToken,
        ...sanitizedUser
      } = user; // be careful, you need to omit other private attributes yourself
      return sanitizedUser;
    };
  
    plugin.controllers.user.me = async (ctx) => {
      if (!ctx.state.user) {
        return ctx.unauthorized();
      }
      const pop = ctx.query.populate
      const user = await strapi.entityService.findOne(
        'plugin::users-permissions.user',
        ctx.state.user.id, {
          populate: pop ? pop : ['role' /* extra default populated fields here */]
        }
      );
  
      ctx.body = sanitizeOutput(user);
    };
  
    return plugin;
  };
```

**Comment by bjorelid** [comment]
> I'm not sure if I'am doing well but even after version update to 4.1.10 I can't populate fields with the /api/users/me controller
> 
> e.g.:
> 
> `/api/users/me?[populate][relation]=*` `/api/users/me?relation=*`
> 
> So I put this code based on some code I found somewhere here:
> 
> ```
> // ISSUE: 
> // https://github.com/strapi/strapi/issues/11957
> 
> const Query = require("mysql/lib/protocol/sequences/Query");
> 
> // BUGFIX
> module.exports = plugin => {
>     const sanitizeOutput = (user) => {
>       const {
>         password,
>         resetPasswordToken,
>         confirmationToken,
>         ...sanitizedUser
>       } = user; // be careful, you need to omit other private attributes yourself
>       return sanitizedUser;
>     };
>   
>     plugin.controllers.user.me = async (ctx) => {
>       if (!ctx.state.user) {
>         return ctx.unauthorized();
>       }
>       const pop = ctx.query.populate
>       const user = await strapi.entityService.findOne(
>         'plugin::users-permissions.user',
>         ctx.state.user.id, {
>           populate: pop ? pop : ['role' /* extra default populated fields here */]
>         }
>       );
>   
>       ctx.body = sanitizeOutput(user);
>     };
>   
>     return plugin;
>   };
> ```

I'm experiencing the same problem with version 4.1.12. Populating works on 'api/users' but not on 'api/users/me'.

**Comment by rvachev** [comment]
> > I'm not sure if I'am doing well but even after version update to 4.1.10 I can't populate fields with the /api/users/me controller
> > e.g.:
> > `/api/users/me?[populate][relation]=*` `/api/users/me?relation=*`
> > So I put this code based on some code I found somewhere here:
> > ```
> > // ISSUE: 
> > // https://github.com/strapi/strapi/issues/11957
> > 
> > const Query = require("mysql/lib/protocol/sequences/Query");
> > 
> > // BUGFIX
> > module.exports = plugin => {
> >     const sanitizeOutput = (user) => {
> >       const {
> >         password,
> >         resetPasswordToken,
> >         confirmationToken,
> >         ...sanitizedUser
> >       } = user; // be careful, you need to omit other private attributes yourself
> >       return sanitizedUser;
> >     };
> >   
> >     plugin.controllers.user.me = async (ctx) => {
> >       if (!ctx.state.user) {
> >         return ctx.unauthorized();
> >       }
> >       const pop = ctx.query.populate
> >       const user = await strapi.entityService.findOne(
> >         'plugin::users-permissions.user',
> >         ctx.state.user.id, {
> >           populate: pop ? pop : ['role' /* extra default populated fields here */]
> >         }
> >       );
> >   
> >       ctx.body = sanitizeOutput(user);
> >     };
> >   
> >     return plugin;
> >   };
> > ```
> 
> I'm experiencing the same problem with version 4.1.12. Populating works on 'api/users' but not on 'api/users/me'.

The same with 4.2.0😫

**Comment by Convly** [comment]
Hey folks, I've tried reproducing the issue but everything seems to work fine on the /users/me endpoint with a populate.

Does anyone have a repository or a small video/gif I could look at to try to find the issue? 

image
image
image

**Comment by M1CK431** [comment]
What about the graphql API?

**Comment by rvachev** [comment]
> Hey folks, I've tried reproducing the issue but everything seems to work fine on the /users/me endpoint with a populate.
> 
> Does anyone have a repository or a small video/gif I could look at to try to find the issue?
> 
> image image image

before the strapi version 4.2.2 populate on `/users/me` didn't work

**Comment by Puvvl** [comment]
Seems it had been fixed in 4.2.2
Screenshot 2022-07-27 151730

**Comment by derrickmehaffy** [comment]
Going to lock this as it's pretty much confirmed that it's fixed as of v4.2.2+

If anyone else hits this issue or something like it please open a new bug report.

**Pull request merged**
